src/http/request.cr

```diff
@@ -93,32 +93,119 @@ class HTTP::Request
 
   # :nodoc:
   record BadRequest
+  # :nodoc:
+  record EOF
+  # :nodoc:
+  record RequestLine, method : String, resource : String, http_version : String
 
   # Returns a `HTTP::Request` instance if successfully parsed,
   # `nil` on EOF or `BadRequest` otherwise.
-  def self.from_io(io)
+  def self.from_io(io) : HTTP::Request | BadRequest | Nil
+    line = parse_request_line(io)
+
+    case line
+    when EOF
+      nil
+    when BadRequest
+      BadRequest.new
```

Comment:
Why not simply return `line`? This entire `case` could even be replaced by `return line unless line.is_a?(RequestLine)` when `EOF` would just be `nil` instead of an extra type. Having lesser types is better and I don't see no reason why this couldn't be `nil`.

Reply:
Yes, this was initially `nil` but I thought it was clearer with `EOF` because then I don't have to document that `nil` means `EOF`. I'll try it and see how it looks. After all it's just a small internal method.

Reply:
Fixed!